source/agora/consensus/SlashPolicy.d

```diff
@@ -173,6 +174,12 @@ public class SlashPolicy
                 valid_keys ~= key;
         }
 
+        version (unittest)
+        {
+            if (valid_keys.length == 0)
+                return getTestRandomSeed();
```

Comment:
Why is this necessary?

Reply:
There are so many cases in unit tests where the pre-images needed don't exist. It's because the tests don't in fact need to focus on the existence of pre-images. So I decided to just add codes that work around errors due to missing pre-images because it will take so much time to provide proper preimages in the unit tests.

Reply:
Isn't that what [waitForPreimages](https://github.com/bpfkorea/agora/blob/c7b78c6092a03efe42598348a2cb9b10cdb9e66b/source/agora/test/Base.d#L677) and [expectBlock](https://github.com/bpfkorea/agora/blob/c7b78c6092a03efe42598348a2cb9b10cdb9e66b/source/agora/test/Base.d#L657) which calls it is for?

Reply:
Yes, right. But I mean unit tests, not network tests. The functions are not used in unit tests.

Reply:
I see. The issue I see is that this code will also be used by the networking tests.

Maybe a better way is to provide a custom SlashPolicy in the non-networking tests, which overrides `getRandomSeed`.

Reply:
I think it's better, thanks.